DIPs/DIP1009.md

```diff
@@ -42,38 +44,51 @@ body
 
 ## Description
 
-Allow `in`, `out`, and `invariant` contracts to be written as parenthesized expressions. Allow multiple expressions of this kind. Create a new grammatical construct called ContractParameters, which copies and replaces the parameters to `assert` expressions. Use the new construct for `in`, `out`, `assert`, and `invariant` expressions alike. `in` and `out` expressions must come at the end of the function declarator suffix, and before the regular contracts, if any. `invariant` expressions require a semicolon terminator. (Body-less interface functions which use only the new syntax also require a semicolon terminator.) The keyword `do` or `body` is not required after the new `in` and `out` expressions.  The choice of syntax for `out` expressions requires a separate analysis (see the Analysis section below). For the sake of the following example, the `out` expression syntax requires a semicolon separating an optional return identifier from its subsequent ContractParameters:
+Take the grammar for existing `assert` parameters, and call it ContractParameters. Copy this grammar and use it in all the new syntaxes.
+
+Introduce two new `in` and `out` contract syntaxes, called InExpression and OutExpression.
+
+An InExpression is `in`, followed by parentheses containing ContractParameters, exactly mimicking an `assert` call.
+
+The recommended OutExpression syntax is `out`, followed by parentheses containing an optional return identifier, a manditory semicolon, and ContractParameters (see Analysis below).
 
+For example:
 ```d
 int fun(ref int a, int b)
    in(a > 0)
    in(b >= 0, "b cannot be negative!")
    out(r; r > 0, "return must be positive")
-   out(; a !=0)
+   out(; a != 0)
 {
    // function body
 }
-
-class C {
-   int data;
-   invariant(data != 0, "data cannot be 0");
-}
-
+```
+InExpressions and OutExpressions are considered the last part of the function signature. Multiple such expressions are permitted. The existing grammar for the function body, which includes ordinary bracketed `in` and `out` contracts, is unchanged. Therefore no `do` or `body` keyword is required after a new InExpression or OutExpression, unless it is also followed by a bracketed `in` or `out` contract. In the case of body-less virtual interface functions, a semicolon is required if there is no bracketed contract:
+```d
 interface I {
    int fun(int a) in(a > 0); 
 }
 ```
-In existing D, the above contracts would have to be written as (and would in most cases lower to):
+A new type of `invariant` is introduced, again mimicking an `assert` call:
+```d
+class C {
+   int data;
+   invariant(data != 0, "cannot be 0");
+}
+```
+By default, all the new contract expressions are treated simply as `assert` calls. This DIP does not prohibit other types of checking, but `assert` is expected to be the norm. In existing D, therefore, the above contracts would have to be written as, and would lower to:
```

Comment:
Replace this with:
"Contracts using the new syntax should be lowered by the compiler to the existing syntax, so that the above example `fun` would be lowered to the following:"

Then, after the example:
"This DIP does not prohibit the use of validation constructs other than `assert` inside a contract, but is based on the assumption that `assert` is the norm. Providing a means to replace `assert` in the lowered code with an alternative is beyond the scope of this DIP, but could make for an interesting topic for a future DIP."